test/sanity/code-smell/mypy.py

```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env python
+# Copyright: (c) 2020-2021, Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+"""Making sure the type annotations are correct across packages."""
+
+import subprocess
+import sys
+
+
+def run_mypy(python_version=None):
+    """Execute mypy against a given Python version.
+
+    This function proxies mypy's return code and filters out
+    the output so that `ansible-test` wouldn't get confused.
+    """
+    mypy_cmd = (
+        sys.executable, '-m', 'mypy',
+        '--show-error-codes',
+        *(
+            () if python_version is None
+            else ('--python-version', python_version)
+        ),
+        # 'hacking/shippable/incidental.py',
+        'lib/ansible/galaxy/collection/',
+        'lib/ansible/galaxy/dependency_resolution',
+        # 'test/lib/ansible_test/_internal',
+        # 'test/utils/shippable/check_matrix.py',
+    )
+    try:
+        mypy_out = subprocess.check_output(mypy_cmd, universal_newlines=True)
+    except subprocess.CalledProcessError as proc_err:
+        mypy_out = proc_err.output
+        return_code = proc_err.returncode
+    else:
+        return_code = 0
+
+    if not return_code:
+        # NOTE: Interrupt because `ansible-test sanity` runner treats any
+        # NOTE: output as a linting failure.
+        return return_code
+
+    print(
+        'Results of type checking with mypy against '
+        f'{python_version or "unspecified"!s} Python '
+        f'(return code {return_code!s}):',
+        file=sys.stderr,
+    )
+    for line in mypy_out.splitlines():
+        if line.startswith((u'Success: no issues found in ', u'Found ')):
+            out_stream = sys.stderr
+        else:
+            out_stream = sys.stdout
+
+        print(line, file=out_stream)
+
+    return return_code
+
+
+def main():
+    """Validate type annotations with mypy against Python 2 and 3."""
+    target_pythons = '3.9', '3.6', '2.7'
```

Comment:
Type annotations may differ under different Pythons. The major difference is between Python 2 and Python 3. But different releases of Python 3 may have a different amount of type stubs so it's good to at least include min and max versions of the Python 3 spectrum.

Reply:
The sanity test should run under the version of Python selected by ansible-test. We already have multi-version support for the built-in sanity tests. I can update ansible-test to enable that functionality for code-smell tests.

Reply:
@mattclay that's not possible: mypy only runs under Python 3.5+ runtimes but is able to lint Python 2.7 code while running under Python 3. Ref: https://mypy.readthedocs.io/en/stable/getting_started.html#installing-and-running-mypy.

> Although you must install Python 3 to run mypy, mypy is fully capable of type checking Python 2 code as well: just pass in the `--py2` flag. See [Type checking Python 2 code](https://mypy.readthedocs.io/en/stable/python2.html#python2) for more details.
>
> ```console
> $ mypy --py2 program.py
> ```

Reply:
That shouldn't be a problem, since we're only testing controller-side code which will soon be Python 3.8+ only.

Reply:
> That shouldn't be a problem, since we're only testing controller-side code which will soon be Python 3.8+ only.

I was hoping to get it integrated for 2.11 so that the dependency resolver would be linted in this version and this sort of check would cover any backports.

Reply:
@mattclay why can't we also test the remote side as well?